examples/multigpu_advtrain/target_attack.py

```diff
@@ -0,0 +1,257 @@
+                                'eps_iter': np.float32,
+                                'y': np.float32,
+                                'y_target': np.float32,
+                                'clip_min': np.float32,
+                                'clip_max': np.float32}
+        self.structural_kwargs = ['ord', 'nb_iter']
+
+        if not isinstance(self.model, Model):
+            self.model = MLP_probs(self.model)
+
+    def clip_eta(self, eta):
+        # Clipping perturbation eta to self.ord norm ball
+        if self.ord == np.inf:
+            eta = tf.clip_by_value(eta, -self.eps, self.eps)
+        elif self.ord in [1, 2]:
+            reduc_ind = list(xrange(1, len(eta.get_shape())))
+            if self.ord == 1:
+                norm = tf.reduce_sum(tf.abs(eta),
+                                     reduction_indices=reduc_ind,
+                                     keep_dims=True)
+            elif self.ord == 2:
+                norm = tf.sqrt(tf.reduce_sum(tf.square(eta),
+                                             reduction_indices=reduc_ind,
+                                             keep_dims=True))
+            eta = eta * self.eps / norm
+        return eta
+
+    def PGDpgd(self, x, **kwargs):
+        from cleverhans.utils_tf import model_loss
+
+        eta = tf.random_uniform(tf.shape(x), -self.eps, self.eps)
+        eta = self.clip_eta(eta)
+
+        if self.y is not None:
+            y = self.y
+        else:
+            preds = self.model.get_probs(x)
+            preds_max = tf.reduce_max(preds, 1, keep_dims=True)
+            y = tf.to_float(tf.equal(preds, preds_max))
+            y = y / tf.reduce_sum(y, 1, keep_dims=True)
+        y = tf.stop_gradient(y)
+
+        for i in range(self.nb_iter):
+            adv_x = tf.stop_gradient(x + eta)
+            preds = self.model.get_probs(adv_x)
+            loss = model_loss(y, preds)
+            grad, = tf.gradients(loss, adv_x)
+            scaled_signed_grad = self.eps_iter * tf.sign(grad)
+            adv_x = adv_x + scaled_signed_grad
+            adv_x = tf.clip_by_value(adv_x, self.clip_min, self.clip_max)
+            eta = adv_x - x
+            eta = self.clip_eta(eta)
+
+        adv_x = x + eta
+        if self.clip_min is not None and self.clip_max is not None:
+            adv_x = tf.clip_by_value(adv_x, self.clip_min, self.clip_max)
+        adv_x = tf.stop_gradient(adv_x)
+
+        return adv_x
+
+    def PGDpgd_singleStep(self, x, eta, y):
+        adv_x = tf.stop_gradient(x + eta)
+        preds = self.model.get_probs(adv_x)
+        loss = model_loss(y, preds)
+        grad, = tf.gradients(loss, adv_x)
+        scaled_signed_grad = self.eps_iter * tf.sign(grad)
+        adv_x = adv_x + scaled_signed_grad
+        adv_x = tf.clip_by_value(adv_x, self.clip_min, self.clip_max)
+        eta = adv_x - x
+        eta = self.clip_eta(eta)
+        return x, eta, y
+
+    def PGDpgd_multigpu(self, x, **kwargs):
```

Comment:
for this method name, I still don't know what "PGDpgd" means

Reply:
Fixed in 460d3f3.